Userland/Libraries/LibLocate/Crawler.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2021, Aron Lander <aron@aronlander.se>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibLocate/Crawler.h>
+
+namespace Locate {
+
+Crawler::Crawler(String path)
+{
+    m_dir_queue.enqueue(make<Locate::DirectoryInfo>(Locate::DirectoryInfo {
+        path,
+        m_id_counter,
+        0,
+        SinglyLinkedListWithCount<NonnullOwnPtr<Locate::ChildInfo>>() }));
```

Comment:
You're invoking `DirectoryInfo`'s move ctor here (if there is one, otherwise this is a copy), is there any reason not to do this?
```c++
m_dir_queue.enqueue(make<Locate::DirectoryInfo>(
        path,
        m_id_counter,
        0,
        SinglyLinkedListWithCount<NonnullOwnPtr<Locate::ChildInfo>>()));
```

Reply:
Since DirectoryInfo is a struct, it doesn't (afaik) have constructor. Did btw. try your suggestion and it will indeed not work (due to the lack of a constructor in DirectoryInfo). Since DirectoryInfo is just a container, I thought that it would be overkill to make a complete class out of it.

I did also notice now that I hadn't used the .attribute = value notation, which hurts my eyes (well, rather, the lack of it), so fixed that was well while looking at this.

Reply:
Alright, it's probably gonna be optimised away anyway

Reply:
Still a bit new to smart pointers, so I might have done something stupid here, but the idea was basically only to have the DirectoryInfo stored in one place and thus have it moved when dequeued rather than copied (which is what I understood would happen from when reading up on the workings of smart pointers). I.e. from what I undestood, this should also result in the ownership changing when getting dequeued (and thus, the DirectoryInfo object after being handled in the locate.cpp being cleaned up at the end of the main loop due to going out of scope).

Reply:
I'm not too sure which would be better, the members of this struct look rather cheap to move() out (which is what dequeue should be doing), maybe except the String member which has to go through an atomic op...
I guess it's alright to keep this as-is (so feel free to disregard the question below about `NonnullOwnPtr`)

Reply:
> Alright, it's probably gonna be optimised away anyway

I'd rather redo it properly though :)